core/trino-parser/src/main/java/io/trino/sql/tree/GrantObject.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.tree;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public class GrantObject
+        extends Node
+{
+    private final Optional<String> entityKind;
+    private final QualifiedName name;
+
+    public GrantObject(NodeLocation location, Optional<String> entityKind, QualifiedName name)
+    {
+        super(Optional.of(location));
+        this.entityKind = requireNonNull(entityKind, "entityKind is null");
+        this.name = requireNonNull(name, "name is null");
+    }
+
+    public Optional<String> getEntityKind()
+    {
+        return entityKind;
+    }
+
+    public QualifiedName getName()
+    {
+        return name;
+    }
+
+    @Override
+    public List<? extends Node> getChildren()
+    {
+        return ImmutableList.of();
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        GrantObject that = (GrantObject) o;
+        return Objects.equals(getEntityKind(), that.getEntityKind()) && Objects.equals(getName(), that.getName());
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .add("entityKind", entityKind)
+                .add("qualifiedName", name)
+                .toString();
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(getEntityKind(), getName());
```

Comment:
Use the fields directly instead of going through the public getters, here and in equals(). While not the case today, it's conceivable that the getters might return a derived view of the internal state of the object that doesn't match the internal state.

Reply:
Good point, and a lesson I'll remember in the future.  Changed.